src/api/taskRuns.js

```diff
@@ -76,73 +76,141 @@ export function cancelTaskRun({ name, namespace }) {
 }
 
-export function createTaskRun({
+export function getTaskRunPayload({
   kind,
   labels,
   namespace,
   nodeSelector,
   params,
-  serviceAccount,
   taskName,
-  taskRunName = `${taskName}-run-${Date.now()}`,
+  taskRunName = `${taskName ? `${taskName}-run` : 'run'}-${Date.now()}`,
+  serviceAccount,
   timeout
 }) {
   const payload = {
     apiVersion: `tekton.dev/${getTektonPipelinesAPIVersion()}`,
     kind: 'TaskRun',
     metadata: {
       name: taskRunName,
-      namespace,
-      labels
+      namespace
     },
     spec: {
-      params: [],
       taskRef: {
         name: taskName,
         kind: kind || 'Task'
       }
     }
   };
-  if (nodeSelector) {
-    payload.spec.podTemplate = { nodeSelector };
+  if (labels) {
+    payload.metadata.labels = labels;
   }
   if (params) {
     payload.spec.params = Object.keys(params).map(name => ({
       name,
       value: params[name]
     }));
   }
+  if (nodeSelector) {
+    payload.spec.podTemplate = { nodeSelector };
+  }
   if (serviceAccount) {
     payload.spec.serviceAccountName = serviceAccount;
   }
   if (timeout) {
     payload.spec.timeout = timeout;
   }
+
+  return payload;
+}
+
+export function createTaskRun({
+  kind,
+  labels,
+  namespace,
+  nodeSelector,
+  params,
+  serviceAccount,
+  taskName,
+  taskRunName = `${taskName}-run-${Date.now()}`,
+  timeout
+}) {
+  const payload = getTaskRunPayload({
+    kind,
+    labels,
+    namespace,
+    nodeSelector,
+    params,
+    serviceAccount,
+    taskName,
+    taskRunName,
+    timeout
+  });
   const uri = getTektonAPI('taskruns', { namespace });
   return post(uri, payload).then(({ body }) => body);
 }
 
-export function rerunTaskRun(taskRun) {
-  const { annotations, labels, name, namespace } = taskRun.metadata;
+export function createTaskRunRaw({ namespace, payload }) {
+  const uri = getTektonAPI('taskruns', { namespace });
+  return post(uri, payload).then(({ body }) => body);
+}
+
+export function generateNewTaskRunPayload({ taskRun, rerun }) {
+  const { annotations, labels, name, namespace, generateName } =
+    taskRun.metadata;
 
   const payload = deepClone(taskRun);
   payload.apiVersion =
     payload.apiVersion || `tekton.dev/${getTektonPipelinesAPIVersion()}`;
   payload.kind = payload.kind || 'TaskRun';
+
+  function getGenerateName() {
+    if (rerun) {
+      return getGenerateNamePrefixForRerun(name);
+    }
+
+    return generateName || `${name}-`;
+  }
+
   payload.metadata = {
-    annotations,
-    generateName: getGenerateNamePrefixForRerun(name),
-    labels: {
-      ...labels,
-      'dashboard.tekton.dev/rerunOf': name
-    },
+    annotations: annotations || {},
+    generateName: getGenerateName(),
+    labels: labels || {},
     namespace
   };
+  if (rerun) {
+    payload.metadata.labels['dashboard.tekton.dev/rerunOf'] = name;
+  }
 
-  delete payload.metadata.labels['tekton.dev/task'];
+  Object.keys(payload.metadata.labels).forEach(label => {
```

Comment:
This clean up is also done as part of pipelineRun API: https://github.com/tektoncd/dashboard/blob/63dce0847d4de489229360b5f32c777508f5b7c4/src/api/pipelineRuns.js#L201

Candidate for making it a util function to remove the `tekton.dev` labels from metadata?

Reply:
done